### PR TITLE
Parametrise path scheme in `dataset_path`

### PR DESCRIPTION
**Pull request by vladsavelyev**
Parametrise the cloud provider path scheme for `dataset_path`. For Azure, assuming the Hail's `hail-az` scheme.

Added: 
* `cloudpath_hail_az.py` module is an extension for `cloudpathlib` to support the `hail-az` scheme with `CloudPath`.
* Into `hail_bactch.py`: 
   * An abstract class for path protocol along with 2 implementations for `gs` and `hail-az` correspondingly.
   * Default path scheme is `gs`, and it can be parametrise in the config through `workflow/path_scheme`.
   * Explicitly parametrise `dataset_path()` with `dataset`, `access_level`, and `path_scheme`.
   * Add `Path` and `to_path` shortcuts are used everywhere in `cpg-pipes`, would be nice to have them in `cpg-utils` to aid moving code from one repository into another.

### CODE REVIEW

File: cpg_utils/hail_batch.py [hunk 1]
```diff
@@ -64,6 +72,9 @@ def init_batch(**kwargs):
     kwargs : keyword arguments
         Forwarded directly to `hl.init_batch`.
     """
+
+    billing_project = os.getenv('HAIL_BILLING_PROJECT')
```
Comment (lgruen) [inline cpg_utils/hail_batch.py]: Looks like this didn't get resolved properly when merging with HEAD?

**Review by lgruen — CHANGES_REQUESTED** [review]

File: cpg_utils/hail_batch.py (hunk 1 above)
Comment (vladsavelyev) [reply cpg_utils/hail_batch.py]: Oh. Yeah, merge artefact. Fixed.

Comment (vladsavelyev) [reply cpg_utils/hail_batch.py]: There are few more merge artefacts, will address them now

File: cpg_utils/cloudpath_hail_az.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    ):
+        if isinstance(cloud_path, str):
+            parsed = urlparse(cloud_path)
+            m = re.match(
+                r'(?P<account>[a-z0-9]+)(\.(?P<type>blob|dfs)(\.core\.windows\.net)?)?',
+                parsed.netloc,
+                flags=re.IGNORECASE,
+            )
+            if m is None:
+                raise ValueError(f"Bad Azure path '{cloud_path}'")
```
Comment (lgruen) [inline cpg_utils/cloudpath_hail_az.py]:
```suggestion
                raise ValueError(f'Bad Azure path "{cloud_path}"')
```

File: cpg_utils/hail_batch.py [hunk 3]
```diff
@@ -81,11 +54,21 @@ def copy_common_env(job: hb.batch.job.Job) -> None:
...
     The environment variable values are extracted from the current process and
-    copied to the environment dictionary of the given Hail Batch job.
-    """
-    # If possible, please don't add new environment variables here, but instead add
-    # config variables.
-    for key in ('CPG_CONFIG_PATH',):
+    copied to the environment dictionary of the given Hail Batch job."""
+
+    for key in (
+        'CPG_ACCESS_LEVEL',
```
Comment (lgruen) [inline cpg_utils/hail_batch.py]: Looks like there's still some merge remnants.

**Review by lgruen — CHANGES_REQUESTED** [review]

File: cpg_utils/hail_batch.py (hunk 3 above)
Comment (vladsavelyev) [reply cpg_utils/hail_batch.py]: Somehow screwed up merge with main. Should be good now!

File: cpg_utils/hail_batch.py [hunk 4]
```diff
@@ -101,7 +117,73 @@ def remote_tmpdir(hail_bucket: Optional[str] = None) -> str:
...
+        """Build full path from prefix and suffix"""
+
+    @staticmethod
+    def parse(val: str) -> 'PathScheme':
+        """Parse subclass name from string"""
+        if val == 'gs':
+            return GSPathScheme()
+        if val in ['az', 'hail-az']:
+            return AzurePathScheme()
+        raise ValueError(f'Unsupported path format: {val}. Available: gs, hail-az')
```
Comment (lgruen) [inline cpg_utils/hail_batch.py]:
```suggestion
        match val:
            case 'gs':
                return GSPathScheme()
            case 'hail-az':
                return AzurePathScheme()
            case _:
                raise ValueError(f'Unsupported path format: {val}. Available: gs, hail-az')
```

File: cpg_utils/hail_batch.py [hunk 5]
```diff
@@ -101,7 +117,73 @@ def remote_tmpdir(hail_bucket: Optional[str] = None) -> str:
...
+
+class AzurePathScheme(PathScheme):
+    """
+    Azure Blob Storage path scheme, following the Hail Batch hail-az format.
+    """
+
+    def __init__(self, account: Optional[str] = 'cpg'):
+        config = get_config()
+        self.scheme = 'hail-az'
+        self.account = config['workflow'].get('azure_account', account)
```
Comment (lgruen) [inline cpg_utils/hail_batch.py:171]:
```suggestion
        self.scheme = 'hail-az'
        self.account = get_config()['workflow'].get('azure_account', account)
```

**Review by lgruen — APPROVED** [review]

File: cpg_utils/hail_batch.py (hunk 5 above)
Comment (vladsavelyev) [reply cpg_utils/hail_batch.py:171]: Any reason why it's better?

File: cpg_utils/cloudpath_hail_az.py (hunk 2 above)
Comment (vladsavelyev) [reply cpg_utils/cloudpath_hail_az.py]: Thanks!

File: cpg_utils/hail_batch.py (hunk 5 above)
Comment (lgruen) [reply cpg_utils/hail_batch.py:171]: Just that `config` isn't reused anywhere, so it's just one less variable to keep track of :)

**Pull request merged**